akka-http-core/src/main/scala/akka/http/scaladsl/model/HttpEntity.scala

```diff
@@ -605,19 +605,21 @@ object HttpEntity {
           // "no limit"
           case Some(SizeLimit(bytes, cl @ Some(contentLength))) ⇒
             if (contentLength > bytes) throw EntityStreamSizeException(bytes, cl)
-          // else we still count but never throw an error
+            maxBytes = bytes
+            bytesLeft = bytes
           case Some(SizeLimit(bytes, None)) ⇒
             maxBytes = bytes
             bytesLeft = bytes
           case None ⇒
+          // else we still count but never throw an error
         }
       }
 
       override def onPush(): Unit = {
         val elem = grab(in)
         bytesLeft -= sizeOf(elem)
         if (bytesLeft >= 0) push(out, elem)
-        else failStage(EntityStreamSizeException(maxBytes))
+        else failStage(EntityStreamSizeException(maxBytes, Some(maxBytes - bytesLeft)))
```

Comment:
actually, I guess we should **never** failStage when something on our side misbehaves and instead start a graceful shutdown with an error (maybe `INTERNAL_ERROR`?). or better just push and let the client give us a `FRAME_SIZE_ERROR`. not sure what's better